File: js_algo/leetcode/146_LRU_cache.js
// https://leetcode.com/problems/lru-cache/

const Node = function (key, value) {
    this.value = value
    this.key = key
    this.next = null
    this.prev = null
}

class DoublyLinkedList {
    constructor() {
        this.head = null
        this.tail = null
        this.length = 0
    }

    clearList() {
        this.tail = null
        this.head = null
        this.length = 0
    }

    push(newNode) {
        if (this.length) {
            this.tail.next = newNode
            newNode.prev = this.tail
            this.tail = newNode
        } else {
            this.head = newNode
            this.tail = newNode
        }

        this.length += 1

        return this
    }

    pop() {
        if (!this.length) {
            return
        }
        const removedNode = this.tail

        if (this.length === 1) {
            this.clearList()
        } else {
            this.tail = removedNode.prev
            this.tail.next = null
            this.length -= 1
        }

        return removedNode
    }

    shift() {
        if (!this.length) {
            return
        }

        const removedNode = this.head

        if (this.length === 1) {
            this.clearList()
        } else {
            this.head = removedNode.next
            this.length -= 1
        }

        return removedNode
    }

    get(index) {
        if (index >= this.length || index < 0) {
            return null
        }

        let result = this.head
        let i = 0
        while (result !== null) {
            if (i === index) {
                break
            }

            if (result.next) {
                result = result.next
            }
            i++
        }

        return result
    }

    set(index, value) {
        if (index > this.length || index < 0) {
            return false
        }

        if (index === this.length) {
            this.push(value)
            return true
        }

        const node = this.get(index)

        if (node) {
            node.value = value
            return true
        }

        return false
    }

    remove(index) {
        if (index > this.length) {
            return
        }

        if (index === 0) {
            this.shift()
            return
        }

        if (index === this.length) {
            this.pop()
            return
        }

        const node = this.get(index)

        if (node) {
            node.prev.next = node.next
            node.next.prev = node.prev

            this.length -= 1
        }
    }

    swapTail(node) {
        if (this.length <= 1) {
            return null
        }

        const { prev, next } = node

		if (node === this.head) {
			this.head = next
			this.head.next = node
		}

		if (node === this.tail) {
			return
		}

		if (prev) {
			prev.next = next
		}

		if (next) {
			next.prev = prev
		}

        this.tail.next = node
        node.prev = this.tail
        this.tail = node
        this.tail.next = null
    }
}

const LRUCache = function (capacity) {
    this.capacity = capacity
    this.storage = {}
    this.dll = new DoublyLinkedList()
}

LRUCache.prototype.get = function (key) {
    const result = this.storage[key]

    if (result === undefined) {
        return -1
    }

    // update LRU order?
	this.dll.swapTail(result)

    return result.value
}

LRUCache.prototype.put = function (key, value) {
    let node = this.storage[key]

    if (node) {
        // check LRU order
        node.value = value

        this.dll.swapTail(node)
        return null
    }

    if (this.dll.length === this.capacity) {
        // remove LRU item && append new key
        const lruItem = this.dll.head
        delete this.storage[lruItem.key]
        this.dll.shift()
    }

    node = new Node(key, value)
    this.storage[key] = node

    this.dll.push(node)
}

const run = () => {
// 	["LRUCache","put","put","get","get","put","get","get","get"]
//   [[2],[2,1],[3,2],[3],[2],[4,3],[2],[3],[4]]
    const lRUCache = new LRUCache(2)

    let result = lRUCache.put(2, 1) // cache is {1=1}
    result = lRUCache.put(3, 1) // cache is {1=1, 2=2}
    result = lRUCache.get(3) // return 1
    result = lRUCache.get(2) // returns -1 (not found)
    result = lRUCache.put(4, 3) // LRU key was 2, evicts key 2, cache is {1=1, 3=3}
    result = lRUCache.get(2) // return -1 (not found)
    result = lRUCache.get(3) // return 3
    result = lRUCache.get(4) // return 4
    // const result = canConstruct('a', 'b')
    console.log(result)
}

export default run

File: python_algo/leetcode/_146_LRU_cache.py
# https://leetcode.com/problems/lru-cache/

class Node:
	def __init__(self, key, value) -> None:
		self.value = value
		self.key = key
		self.next = None
		self.prev = None


class DoublyLinkedList:
	def __init__(self):
		self.head = None
		self.tail = None
		self.length = 0

	def clearList(self):
		self.tail = None
		self.head = None
		self.length = 0

	def push(self, newNode):
		if self.length:
			self.tail.next = newNode
			newNode.prev = self.tail
			self.tail = newNode
		else:
			self.head = newNode
			self.tail = newNode

		self.length += 1

	def pop(self):
		if not self.length:
			return

		removedNode = self.tail

		if self.length == 1:
			self.clearList()
		else:
			self.tail = removedNode.prev
			self.tail.next = None
			self.length -= 1

		return removedNode

	def shift(self):
		if not self.length:
			return

		removedNode = self.head

		if self.length == 1:
			self.clearList()
		else:
			self.head = removedNode.next
			self.length -= 1

		return removedNode

	def swapTail(self, node):
		if self.length <= 1:
			return None

		prev = node.prev
		next = node.next

		if node == self.tail:
			return

		if node == self.head:
			self.head = next
			self.head.prev = None
		else:
			if prev:
				prev.next = next

			if next:
				next.prev = prev

		self.tail.next = node
		node.prev = self.tail
		self.tail = node
		self.tail.next = None


class LRUCache:
	def __init__(self, capacity):
		self.capacity = capacity
		self.storage = {}
		self.dll = DoublyLinkedList()

	def get(self, key):
		result = self.storage.get(key, None)

		if result == None:
			return -1

		self.dll.swapTail(result)

		return result.value


	def put(self, key, value):
		node = self.storage.get(key, None)

		if node:
			# check LRU order
			node.value = value

			self.dll.swapTail(node)
			return None

		if self.dll.length == self.capacity:
			# remove LRU item && append new key
			lruItem = self.dll.head
			self.storage.pop(lruItem.key)
			self.dll.shift()

		node = Node(key, value)
		self.storage[key] = node

		self.dll.push(node)
